README
NAME
    Log::Log4perl::Filter::CallerMatch - Filter Log4perl messages based on
    call frames

VERSION
    version 1.101060

DESCRIPTION
    This Log4perl custom filter checks the call stack using caller() and
    filters the subroutine and package using user-provided regular
    expressions. You can specify a specific call frame to test against, or
    have the filter iterate through a range of call frames.

ATTRIBUTES
  StringToMatch
    A perl5 regular expression, matched against the log message.

  AcceptOnMatch
    Defines if the filter is supposed to pass or block the message on a
    match ("true" or "false").

  PackageToMatch
    A perl5 regular expression, matched against the 1st item in the array
    returned by caller() (e.g. "package")

  SubToMatch
    A perl5 regular expression, matched against the 4th item in the array
    returned by caller() (e.g. "subroutine")

  CallFrame
    The call frame to use when requesting information from caller(). (e.g.
    $i in caller($i)

  MinCallFrame
    The first call frame tested against when iterating through a series of
    call frames. Ignored if CallFrame specified.

  MaxCallFrame
    The last call frame tested against when iterating through a series of
    call frames. Ignored if CallFrame specified.

METHODS
  new
    Constructor. Refer to Log::Log4perl::Filter for more information

  ok
    Decides whether log message should be accepted or not. Refer to
    Log::Log4perl::Filter for more information

USAGE
     # log.conf
     log4perl.logger = ALL, A1
     log4perl.appender.A1        = Log::Log4perl::Appender::TestBuffer
     log4perl.appender.A1.Filter = MyFilter
     log4perl.appender.A1.layout = Log::Log4perl::Layout::SimpleLayout
    
     log4perl.filter.MyFilter                = Log::Log4perl::Filter::CallerMatch
     log4perl.filter.MyFilter.SubToMatch     = WebGUI::Session::ErrorHandler
     log4perl.filter.MyFilter.PackageToMatch = Flux::
     log4perl.filter.MyFilter.StringToMatch  = Operand1

SEE ALSO
    Log::Log4perl::Filter, Log::Log4perl::Filter::StringMatch,
    Log::Log4perl::Filter::LevelMatch, Log::Log4perl::Filter::LevelRange,
    Log::Log4perl::Filter::Boolean

AUTHOR
      Patrick Donelan <pat@patspam.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Patrick Donelan.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Log-Log4perl-Filter-CallerMatch

1.101060  2010-04-16 16:52:00 EST5EDT
    Improved tests

1.100830  2010-03-24 14:48:48 EST5EDT
    Switched to Dist::Zilla
        
0.01    Date/time
    First version, released on an unsuspecting world.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;

my %module_build_args = (
                       'dist_abstract' => 'Filter Log4perl messages based on call frames',
                       'script_files' => [],
                       'license' => 'perl',
                       'build_requires' => {
                                             'Test::More' => '0',
                                             'File::Find' => '0',
                                             'File::Temp' => '0',
                                             'Module::Build' => '0.3601',
                                             'Log::Log4perl' => '0'
                                           },
                       'dist_name' => 'Log-Log4perl-Filter-CallerMatch',
                       'dist_author' => [
                                          'Patrick Donelan <pat@patspam.com>'
                                        ],
                       'module_name' => 'Log::Log4perl::Filter::CallerMatch',
                       'requires' => {
                                       'Log::Log4perl::Filter' => '0',
                                       'perl' => '5.006',
                                       'Log::Log4perl::Config' => '0',
                                       'Carp' => '0'
                                     },
                       'recommends' => {},
                       'configure_requires' => {
                                                 'ExtUtils::MakeMaker' => '6.31',
                                                 'Module::Build' => '0.3601'
                                               },
                       'dist_version' => '1.101060'
                     );


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Log/Log4perl/Filter/CallerMatch.pm
# 
# This file is part of Log-Log4perl-Filter-CallerMatch
# 
# This software is copyright (c) 2010 by Patrick Donelan.
# 
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
# 
package Log::Log4perl::Filter::CallerMatch;
BEGIN {
  $Log::Log4perl::Filter::CallerMatch::VERSION = '1.101060';
}

# ABSTRACT:  Filter Log4perl messages based on call frames

use 5.006;
use strict;
use Log::Log4perl::Config;
use base 'Log::Log4perl::Filter';
use Carp;


sub new {
    my ( $class, %options ) = @_;

    my $self = {
        AcceptOnMatch => 1,
        MinCallFrame  => 0,
        MaxCallFrame  => 5,
        %options,
    };

    $self->{AcceptOnMatch}  = Log::Log4perl::Config::boolean_to_perlish( $self->{AcceptOnMatch} );
    $self->{SubToMatch}     = defined $self->{SubToMatch} ? qr($self->{SubToMatch}) : qr/.*/;
    $self->{PackageToMatch} = defined $self->{PackageToMatch} ? qr($self->{PackageToMatch}) : qr/.*/;
    $self->{StringToMatch}  = defined $self->{StringToMatch} ? qr($self->{StringToMatch}) : qr/.*/;

    if ( defined $self->{CallFrame} ) {
        $self->{MinCallFrame} = $self->{MaxCallFrame} = $self->{CallFrame};
    }

    bless $self, $class;

    return $self;
}


sub ok {
    my ( $self, %p ) = @_;

    my $message = join $Log::Log4perl::JOIN_MSG_ARRAY_CHAR, @{ $p{message} };

    my ( $s_regex, $p_regex, $m_regex ) = ( $self->{SubToMatch}, $self->{PackageToMatch}, $self->{StringToMatch} );
    
    foreach my $i ( $self->{MinCallFrame} .. $self->{MaxCallFrame} ) {
        my ( $package, $sub ) = ( caller $i )[ 0, 3 ];
        # next unless $package;
        # next unless $sub;
        # warn "$package - $sub";
        no warnings;
        if ( $sub =~ $s_regex && $package =~ $p_regex && $message =~ $m_regex ) {
            return $self->{AcceptOnMatch};
        }
    }
    return !$self->{AcceptOnMatch};
}

1;


=pod

=head1 NAME

Log::Log4perl::Filter::CallerMatch - Filter Log4perl messages based on call frames

=head1 VERSION

version 1.101060

=head1 DESCRIPTION

This Log4perl custom filter checks the call stack using caller() and filters
the subroutine and package using user-provided regular expressions. You can specify
a specific call frame to test against, or have the filter iterate through a range of call frames.

=head1 ATTRIBUTES

=head2 StringToMatch

A perl5 regular expression, matched against the log message.

=head2 AcceptOnMatch

Defines if the filter is supposed to pass or block the message on a match (C<true> or C<false>).

=head2 PackageToMatch

A perl5 regular expression, matched against the 1st item in the array returned by caller() (e.g. "package")

=head2 SubToMatch

A perl5 regular expression, matched against the 4th item in the array returned by caller() (e.g. "subroutine")

=head2 CallFrame

The call frame to use when requesting information from caller(). (e.g. $i in caller($i)

=head2 MinCallFrame

The first call frame tested against when iterating through a series of call frames. Ignored if CallFrame specified.

=head2 MaxCallFrame

The last call frame tested against when iterating through a series of call frames. Ignored if CallFrame specified.

=head1 METHODS

=head2 new

Constructor. Refer to L<Log::Log4perl::Filter> for more information

=head2 ok

Decides whether log message should be accepted or not. Refer to L<Log::Log4perl::Filter> for more information

=head1 USAGE

 # log.conf
 log4perl.logger = ALL, A1
 log4perl.appender.A1        = Log::Log4perl::Appender::TestBuffer
 log4perl.appender.A1.Filter = MyFilter
 log4perl.appender.A1.layout = Log::Log4perl::Layout::SimpleLayout
    
 log4perl.filter.MyFilter                = Log::Log4perl::Filter::CallerMatch
 log4perl.filter.MyFilter.SubToMatch     = WebGUI::Session::ErrorHandler
 log4perl.filter.MyFilter.PackageToMatch = Flux::
 log4perl.filter.MyFilter.StringToMatch  = Operand1

=head1 SEE ALSO

L<Log::Log4perl::Filter>,
L<Log::Log4perl::Filter::StringMatch>,
L<Log::Log4perl::Filter::LevelMatch>,
L<Log::Log4perl::Filter::LevelRange>,
L<Log::Log4perl::Filter::Boolean>

=head1 AUTHOR

  Patrick Donelan <pat@patspam.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Patrick Donelan.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__


1;


Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.006; } 

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Log::Log4perl::Filter::CallerMatch',
                       'DISTNAME' => 'Log-Log4perl-Filter-CallerMatch',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.31',
                                                 'Module::Build' => '0.3601'
                                               },
                       'AUTHOR' => 'Patrick Donelan <pat@patspam.com>',
                       'BUILD_REQUIRES' => {
                                             'Test::More' => '0',
                                             'File::Find' => '0',
                                             'File::Temp' => '0',
                                             'Module::Build' => '0.3601',
                                             'Log::Log4perl' => '0'
                                           },
                       'ABSTRACT' => 'Filter Log4perl messages based on call frames',
                       'EXE_FILES' => [],
                       'VERSION' => '1.101060',
                       'PREREQ_PM' => {
                                        'Log::Log4perl::Filter' => '0',
                                        'Log::Log4perl::Config' => '0',
                                        'Carp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{BUILD_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.56) };

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



